### PR TITLE
OCPNODE-2098: Add static pod for kube-rbac-proxy-crio

### PR DESCRIPTION
**Pull request by rphillips**
This PR adds a static pod to the MCO to lay down a kube-rbac-proxy to secure crio's metrics port. The TLS endpoints supports cert based authentication for monitoring.

### Merge information

* Allow static pod tolerations: 4.16/master : Backport to EUS 4.14
* MCO Kube-Rbac-Proxy Static pods 4.16/master : https://github.com/openshift/machine-config-operator/pull/4175 : Backport to EUS 4.14
* Create an upgrade edge after all backports merge

***After all the above backports are merged, then***

* Change the Monitoring Operator to use port 9637 (with client certificates) : https://github.com/openshift/cluster-monitoring-operator/pull/2257 : This PR is only for 4.16

This PR will need to be backported to EUS 4.14.

### CODE REVIEW

**Comment by rphillips** [comment]
/hold cancel

https://github.com/openshift/origin/pull/28636 merged

**Review by harche — COMMENTED** [review]
/lgtm

File: templates/master/01-master-kubelet/_base/files/criometricsproxy.yaml [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+          cpu: 5m
+      containers:
+      - name: kube-rbac-proxy-crio
+        image: {{.Images.kubeRbacProxyImage}}
+        securityContext:
+          privileged: true
+        ports:
+        - containerPort: 9637
+        args:
+        - --secure-listen-address=:9637
```
Comment (cdoern) [inline templates/master/01-master-kubelet/_base/files/criometricsproxy.yaml:56]: any specific reason for this port?

File: templates/worker/01-worker-kubelet/_base/files/criometricsproxy.yaml [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+        ports:
+        - containerPort: 9637
+        args:
+        - --secure-listen-address=:9637
+        - --config-file=/etc/kubernetes/crio-metrics-proxy.cfg
+        - --client-ca-file=/etc/kubernetes/kubelet-ca.crt
+        - --logtostderr=true
+        - --kubeconfig=/var/lib/kubelet/kubeconfig
+        - --tls-cipher-suites=TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256,TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256,TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384,TLS_ECDHE_ECDSA_WITH_AES_256_GCM_SHA384,TLS_ECDHE_RSA_WITH_CHACHA20_POLY1305,TLS_ECDHE_ECDSA_WITH_CHACHA20_POLY1305
+        - --upstream=http://127.0.0.1:9537
```
Comment (cdoern) [inline templates/worker/01-worker-kubelet/_base/files/criometricsproxy.yaml:62]: our kube-rbac-proxy ctrs usually use a different port, I am forgetting which. Should we mimic that here?

File: templates/master/01-master-kubelet/_base/files/criometricsproxy-config.yaml [hunk 3]
```diff
@@ -0,0 +1,11 @@
...
+path: "/etc/kubernetes/crio-metrics-proxy.cfg"
+contents:
+  inline: |-
+    authorization:
+      static:
+        - resourceRequest: false
+          path: /metrics
+          verb: get
+          user:
+            name: system:serviceaccount:openshift-monitoring:prometheus-k8s
```
Comment (cdoern) [inline templates/master/01-master-kubelet/_base/files/criometricsproxy-config.yaml:11]: is this using the openshift-monitoring SA?

In our other kube-rbac-proxy pods we specify which `serviceAccountName:` we want to use. Where is that happening in this PR?

**Review by cdoern — COMMENTED** [review]
some service account and port questions. but looks good!

File: templates/master/01-master-kubelet/_base/files/criometricsproxy.yaml (hunk 1 above)
Comment (rphillips) [reply templates/master/01-master-kubelet/_base/files/criometricsproxy.yaml:56]: The usual port is 9537. To make it easier to identify the secure port I used 9637 (100 more than the insecure one). (9637 is not in use by the /etc/services file).

File: templates/worker/01-worker-kubelet/_base/files/criometricsproxy.yaml (hunk 2 above)
Comment (rphillips) [reply templates/worker/01-worker-kubelet/_base/files/criometricsproxy.yaml:62]: The kube-rbac-proxy is exposing a host network port (9637) and forwarding to the host's 9537 port. I slightly deviated from the usual construct due to host networking.

File: templates/master/01-master-kubelet/_base/files/criometricsproxy-config.yaml (hunk 3 above)
Comment (rphillips) [reply templates/master/01-master-kubelet/_base/files/criometricsproxy-config.yaml:11]: This new kube-rbac-proxy-crio service does not have access to service account secrets due to being a static pod. Instead, we are using client side certificate validation and only allowing the `system:serviceaccount:openshift-monitoring:prometheus-k8s` user.

**Pull request merged**